timers: use only a single TimerWrap instance

apapirovski opened the pull request:
This PR moves almost the full entirety of timers into JS land and hangs them all off one TimerWrap. This simplifies a lot of code related to timer refing/unrefing and significantly improves performance.

```
                                            confidence improvement accuracy (*)    (**)   (***)
timers/timers-breadth.js n=5000000                   *      3.24 %       ±2.84%  ±3.78%  ±4.92%
timers/timers-cancel-pooled.js n=5000000           ***    -16.38 %       ±2.06%  ±2.74%  ±3.57%
timers/timers-cancel-unpooled.js n=1000000         ***     47.81 %       ±1.33%  ±1.79%  ±2.37%
timers/timers-depth.js n=1000                              -0.24 %       ±0.89%  ±1.19%  ±1.56%
timers/timers-insert-pooled.js n=5000000           ***     16.08 %       ±7.09%  ±9.49% ±12.47%
timers/timers-insert-unpooled.js n=1000000         ***    162.84 %       ±4.82%  ±6.49%  ±8.59%
timers/timers-timeout-pooled.js n=10000000         ***     15.90 %       ±1.59%  ±2.11%  ±2.76%
timers/timers-timeout-unpooled.js n=1000000        ***    948.32 %      ±21.44% ±28.89% ±38.36%
```

Fixes: https://github.com/nodejs/node/issues/16105

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14679/
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/181/
CitGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1398/

apapirovski commented:
FWIW browsers just execute whatever timers are due and then run microtasks. There's no consideration to what the durations are. I feel like that's kind of a particularity of our implementation and we should probably just fix the tests.

The only thing I see on CitGM is that one test in coffeescript repl is affected but again, I would rather just submit a PR to fix the test.

apapirovski commented:
To avoid this being `semver-major`, if at all possible, I've re-added the nextTick behaviour back in (along with a corresponding test). We could perhaps discuss that changing in a `semver-major` PR although it's not strictly necessary.

Fishrock123 commented:
I'll try to get to this in the next couple of days and give it a full review.

bnoordhuis commented on test/parallel/test-priority-queue.js:
```diff
@@ -0,0 +1,79 @@
+  assert(queue.remove(5));
+  assert(queue.remove(10));
+  assert(queue.remove(15));
+  const removed = [5, 10, 15];
```
Maybe this?
```js
const removed = [5, 10, 15];
for (const i of removed)
  assert(queue.remove(i));
```

bnoordhuis commented on lib/internal/priority_queue.js:
```diff
@@ -0,0 +1,103 @@
+
+  peek() {
+    const value = this[kHeap][1];
+    return value !== undefined ? value : null;
```
I'm just curious, is there a reason to remap `undefined` to `null`?

apapirovski commented on lib/timers.js:
```diff
@@ -126,36 +122,59 @@ const kRefed = Symbol('refed');
+
+let timerListId = Number.MIN_SAFE_INTEGER;
+let refCount = 0;
+let refStatus = true;
```
It's necessary with `maybeChangeRefStatus` which is generic (so we don't know if the number went up or down). I could instead have `incRef` and `decRef` — then we wouldn't need it. That's probably better in the end, I'll refactor.

apapirovski commented:
> Currently, that's an O(1) operation but with this pull request, it's O(n).

@bnoordhuis It should still be mostly O(1), only O(log n) when it's the last timer in a list and the n here is the number of timer lists so pretty low in most usage.

The architrcture overall is actually almost exactly the same as the current implementation except we use a binary heap implemented in JS over an array rather the timerwrap / libuv one.

Also FWIW the decrease in our clearTimeout benchmark is only 16% but that comes from setting the symbol prop for kRefed status (not the other stuff). But that's also a benchmark where I recently got 1600% bump (the PR landed in 9.x at some point).

apapirovski commented:
@bnoordhuis also, thanks for reviewing! 😊

apapirovski commented:
@nodejs/timers @jasnell @addaleax This is a fairly substantial change to the internals of timers. Would love to get a few more eyes on this, if possible. Thanks in advance!

bnoordhuis commented on test/parallel/test-priority-queue.js:
```diff
@@ -0,0 +1,78 @@
+const PriorityQueue = require('internal/priority_queue');
+
+{
+  // Checks that the queue is fundumentally correct
```
Typo: fundamentally

bnoordhuis commented on benchmark/util/priority-queue.js:
```diff
@@ -0,0 +1,18 @@
+
+function main({ n, type }) {
+  const PriorityQueue = require('internal/priority_queue');
+  const heap = new PriorityQueue();
```
Minor consistency issue: it's called `queue` everywhere else.

apapirovski commented:
@bnoordhuis That's actually exactly what `timers/timers-cancel-unpooled.js` and `timers/timers-insert-unpooled.js` test. Both saw a substantial improvement.

That snippet should be something like `O(n log n)` (not accounting for whatever the cost of managing the Object map is in V8) but that would be the same as now, except there's less overhead because it doesn't have to call into C++ to do it and create a handle.

In the current architecture that snippet would have to create a handle for each timeout list and that would then have to sort it into the libuv binary heap. Then when clearing, it would delete the list and close the libuv timer handle (which would have to remove it from the binary heap).

apapirovski commented:
Actually, I see what you're referencing @bnoordhuis. I think our benchmarks might've misled me here.

bnoordhuis commented:
> based on the assumption that net (and related) are the most significant consumers of timeouts, which means that most apps should only ever have a few timer lists at most

How so?  Socket timers have millisecond resolution and they're updated with every read and write.  I don't think it's uncommon to end up in a situation where there is a list per socket.

Could be mitigated by rounding down to seconds, that makes it 1000x more likely for them to end up in the same list.

apapirovski commented:
> How so? Socket timers have millisecond resolution and they're updated with every read and write. I don't think it's uncommon to end up in a situation where there is a list per socket.

Right but lists are based on the durations not expiries. And the priority queue just tracks expiries for each list.

bnoordhuis commented:
@apapirovski It'd be easier to review if you keep your new changes in a separate commit.

apapirovski commented:
@nodejs/collaborators Could a few more folks have a look, please? This is a major rework of timers and so far only one person has reviewed it. I would truly appreciate a few more eyes.

BridgeAR commented:
@apapirovski are the benchmarks at the top still up to date after the latest changes?

apapirovski commented on benchmark/util/priority-queue.js:
```diff
@@ -0,0 +1,18 @@
+  const queue = new PriorityQueue();
+  bench.start();
+  for (var i = 0; i < n; i++)
+    queue.insert(Math.random() * 1e7 | 0);
```
@BridgeAR based on my testing, with 1e6 items, it ends up playing a very small role. I see extremely tiny variations run to run as a result but instead we'll be more likely to avoid someone over-optimizing this for the benchmark (which is very easy with a priority queue).

lpinca commented:
I think the problem is that most collaborators are not familiar with timers. For example, I can't review for that reason. Performance improvements look great. Did you try to compare the priority queue to userland implementations like https://github.com/lemire/FastPriorityQueue.js to see if we can get even more performance improvements?

apapirovski commented:
> I think the problem is that most collaborators are not familiar with timers. For example, I can't review for that reason. Performance improvements look great.

I understand but even a cursory look is helpful.

> Did you try to compare the priority queue to userland implementations like https://github.com/lemire/FastPriorityQueue.js to see if we can get even more performance improvements?

I haven't seen that before — thank you for linking it! Looks like the algorithms are basically identical. We have the ability for much faster removals at `O(log n)` but we're slightly slower at `shift` because we let the data get garbage collected whereas that implementation just leaves it in the array and lets the user clean it up whenever they want to (which would leak memory in Node.js).

apapirovski commented:
> I'm concerned in the change of semantics with the async_hooks integration. Specifically, this prevents the destroy hook to fire.

It's a single TimerWrap... how could it possibly cause memory to leak?

mcollina commented:
```js
const asyncHook = require('async_hooks').createHook({init, destroy})
const meta = {}
asyncHook.enable()

function init (asyncId, type, triggerAsyncId, resource) {
   meta[asyncId] = { something: true })
}

function destroy (asyncId) {
  delete meta[asyncId]
}
```

In case of TIMERWRAP, only `init()` will be called, or at least this is what the unit tests says. I see that https://github.com/elastic/apm-agent-nodejs/blob/master/lib/instrumentation/async-hooks.js is skipping TIMERWRAP, so we generically we might be ok.

Is the asyncId changing or staying the same?

AndreasMadsen commented:
@mcollina I wouldn't worry about `async_hooks`. The TIMERWRAP is already useless in terms of async context. If is using `async_hooks` for async context and depending on TIMERWRAP they have a bug in their logic.

mcollina commented:
What is TIMERWRAP there for?

mcollina commented:
It definitely does. +1 to the followup PR!

Fishrock123 commented:
I'd still love to give this a review but my life has been exceedingly busy recently, I'm sorry ... hopefully really soon

apapirovski commented on lib/timers.js:
```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
-      } else if (destroyHooksExist() && !timer._destroyed) {
-        emitDestroy(timerAsyncId);
+    if (timer._repeat) {
+      rearm(timer, start);
```
Yeah, it should be. `rearm` won't trigger any user code or async hook events.

apapirovski commented on lib/internal/priority_queue.js:
```diff
@@ -0,0 +1,111 @@
+    const heap = this[kHeap];
+    const value = heap[1];
+    if (value === undefined)
+      return value;
```
I did but maybe it's just more confusing than helpful?

BridgeAR replied:
To me it is more confusing because I first have to check that it is actually undefined in this case but I do not have a strong opinion on it. So it's your call.

apapirovski commented on lib/timers.js:
```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
+  else
+    handle.unref();
+
+  if (list !== undefined) {
```
The while loop above continues either until its empty (no more timeouts on our process) or the next timer list has an expiry in the future. I wasn't clear if you were asking about the outside of the if or the inside, but both are possible.

Fishrock123 commented:
This is semver-major because of (at least) the removal of a potential `{timer}._handle`. I've seen code which uses that.

apapirovski commented:
@Fishrock123 but it's already only on certain timers... ? Are you saying that there's code which calls `unref` and then expects that `_handle` will exist?

Fishrock123 commented:
Pretty sure that's what I've seen - this unrefed stuff _is_ pretty long lived by now. The disappearance of handles in general also causes some issues for certain diagnostic tools but less in general than code which directly hits the timer. (Still reviewing btw)

Fishrock123 commented on lib/timers.js:
```diff
@@ -126,36 +122,72 @@ const kRefed = Symbol('refed');
+// in comparison to other alternative timers architectures.
 
 
 // Object maps containing linked lists of timers, keyed and sorted by their
```
Object map singular now

Fishrock123 commented on lib/timers.js:
```diff
@@ -126,36 +122,72 @@ const kRefed = Symbol('refed');
+
+// This is a priority queue with a custom sorting function that first compares
+// the expiry times of two lists and if they're the same then compares their
+// individual IDs to determine which list was created first.
```
It seems to me if you have two lists of the same duration then something went very wrong?

How would you tell which list to set a timer of 30ms if there were two lists of 30ms?

apapirovski replied:
It's different durations, same expiries. One list is for 120s, the other for 60s but started 60s later.

Fishrock123 replied:
Would tracking the expiry be better? You should be able to check `list[0].expiry` I think?

apapirovski replied:
Well, expiry is the first point of comparison, right? So we need a secondary criteria which tells us which timer is newer. This also happens to exactly match how this works currently.

Fishrock123 replied:
Oh, for some reason I thought it would be comparing `msecs`. Guess that works for me.

Fishrock123 commented on lib/timers.js:
```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
 // TimerWrap backed list if one does not already exist for the specified timeout
 // duration.
-function insert(item, unrefed, start) {
+function insert(item, refed, start) {
```
Why was this all changed from setting unrefed properties to setting refed properties?

Self-edit from later: Ok I guess this can make sense with the concept of a "refCount"... I still wish the initial setting stayed the same, personally.

Fishrock123 commented on lib/timers.js:
```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
   if (list === undefined) {
     debug('no %d list was found in insert, creating a new one', msecs);
-    lists[msecs] = list = new TimersList(msecs, unrefed);
+    lists[msecs] = list = new TimersList(start, msecs);
```
If you stent this `expiry` instead you'd have to do a tad less math in the lines that follow (mostly code clarity)

Fishrock123 commented on lib/timers.js:
```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
+    if (list.expiry > now)
+      break;
+    if (ran)
+      runNextTicks();
```
This won't cause problems if you schedule short timers in next ticks and the existing queues take longer than that amount of time to process?

This is also probably worth leaving a comment for since we usually don't manually run nextTicks.

apapirovski replied:
Sure but we can just come back on the next iteration of the event loop. The current behaviour is pretty problematic because it's possible to just get stuck in timers literally looping forever if nextTicks take too long.

Fishrock123 replied:
Currently timers get called repeatedly but the event loop still turns as far as I'm aware. Here you might loop forever in JS, no?

apapirovski replied:
Current behaviour is... complicated but basically if you run a list, then add a timeout in next tick, then that next tick takes a long time to follow up, then that list will run again, and then you could keep doing all that over and over without a stop.

With the new behaviour, when we call in to execute timers, we only run all timers that were due at that exact time. The concept of `now` never updates. So any timers that weren't due during that loop's tick will wait until the next one.

Fishrock123 replied:
> Current behaviour is...

Hmmm, I thought we had fixed something like that but perhaps not _that_.

> With the new behaviour, ... The concept of `now` never updates

Ehhh, I just hope that doesn't cause any userland ordering problems.

Fishrock123 commented on lib/timers.js:
```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
-      }
-      handle.start(timeRemaining);
+      list.expiry = timer._idleStart + msecs;
+      list.id = timerListId++;
```
Is Id being used as an alternate solution for `Check if this loop iteration is too early for the next timer.`? 

(That doesn't make sense to me how that would work, since you can't have two lists of the same duration anyways or insert will break.)

apapirovski replied:
You could start a timer for 120s and in 60s start a timer for 60s. Both have the same expiry. (This matches what libuv does to keep ordering consistent.)

Fishrock123 commented on lib/timers.js:
```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
+
+  if (list !== undefined) {
+    nextExpiry = list.expiry;
+    handle.start(Math.max(nextExpiry - TimerWrap.now(), 1));
```
Could you add a comment explaining this?

Fishrock123 commented on lib/timers.js:
```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
+      timer[kRefed] = null;
+
+      if (destroyHooksExist() && !timer._destroyed) {
+        emitDestroy(timer[async_id_symbol]);
```
This is wrong, since the `after` callback from `listOnTimeout` will be emitted later than the `destroy` from here.

apapirovski replied:
Destroy is emitted in an async manner. (See next tick for similar scenario.)

Fishrock123 replied:
Ok... but it seems weird that the testcases for this would then have changed.

apapirovski replied:
Only the TimerWrap test cases have changed as far as async hooks go. The ones for Timeouts remain the same.

Fishrock123 commented on lib/timers.js:
```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
+  L.remove(item);
+
+  // We only delete refed lists because unrefed ones are incredibly likely
+  // to come from http and be recreated shortly after.
```
Isn't this a memory leak?

Also, in most situations where pref matters (e.g. under load) won't enough connections be present that timeouts will usually overlap, causing the list to stay alive anyways?

apapirovski replied:
This just keeps existing behaviour. See `reuse`.

Fishrock123 replied:
Right. Ok, I recall a thing about that - we keep the unrefed list "alive" and let libuv clean it up, right?

We might want to test that it is indeed cleaned up when insert is run when that list should be hit (it should be cleaned up at that point).

apapirovski replied:
That's correct, yeah.

> We might want to test that it is indeed cleaned up when insert is run when that list should be hit (it should be cleaned up at that point).

It does work this way. I suppose we could create a test but it would probably necessitate exporting more stuff from internal timers.

Fishrock123 commented on lib/timers.js:
```diff
@@ -430,36 +471,25 @@ function ontimeout(timer) {
-  }
+  const now = TimerWrap.now();
+  if (start !== undefined && (now - start) >= msecs)
+    insert(timer, timer[kRefed], now - msecs + 1);
```
Please keep the comment (and / or elaborate on it).

apapirovski replied:
While writing a more detailed comment for this I realized it's no longer necessary. Since we now always schedule the next run for at least 1ms, we can set the `start` to the correct value in every situation. This is also less likely to cause ordering inconsistencies.

Fishrock123 commented on src/timer_wrap.cc:
```diff
@@ -62,7 +62,6 @@ class TimerWrap : public HandleWrap {
     env->SetProtoMethod(constructor, "hasRef", HandleWrap::HasRef);
 
     env->SetProtoMethod(constructor, "start", Start);
-    env->SetProtoMethod(constructor, "stop", Stop);
```
I'm sure I've seen timer handles be used by modules in the wild, I wonder if the impact of this might be more problematic than it would seem at first glance?

Technically not "stable api", but this has existed pretty much forever.

apapirovski replied:
I did a scan a didn't find a single use of it. We've also made TimerWrap completely impossible to use by anything but our own timers because of the way OnTimeout works.

Fishrock123 replied:
Oh, right. 😕

apapirovski replied:
FWIW given that nobody complained when that happened, I think TimerWrap likely has 0 usage out there... :)

Fishrock123 commented on test/async-hooks/test-timerwrap.setInterval.js:
```diff
@@ -51,6 +51,6 @@ function onexit() {
   hooks.sanityCheck('TIMERWRAP');
 
-  checkInvocations(t, { init: 1, before: 3, after: 3, destroy: 1 },
+  checkInvocations(t, { init: 1, before: 3, after: 3 },
```
This seems wrong. Is it from the async-hooks place I noted previously?

apapirovski replied:
There's a single TimerWrap, there's no real opportunity to destroy it before exiting.

Fishrock123 replied:
Won't something with a `uv_before` handle work? I think an immediate should do the trick?

apapirovski replied:
> I'm still a little confused as to why handle cleanup doesn't happen this turn of the event loop, though?

There's one global handle for the duration of the process. The goal is to eventually have this be an Environment bound uv_timer_t much like Immediates, rather than a HandleWrap. (See my other comment.)

By the time the process is exiting, it's too late to cleanup handles because they take one event loop run to fully clear plus destroy hooks won't be emitted anyway because they run in an immediate.

apapirovski replied:
But there's nothing to delay for in this case. The handle doesn't get destroyed during the duration of the process. It starts on the first timer run and then exists until the process exits.

(This test is going away when TimerWrap goes.)

apapirovski replied:
Feel free to have a look at for the follow-up that removes TimerWrap altogether: https://github.com/apapirovski/node/commit/d8f075da7b8129f59b1d537b6475394425b81c3e

Fishrock123 commented on test/async-hooks/test-timerwrap.setTimeout.js:
```diff
@@ -69,8 +54,6 @@ function onexit() {
   hooks.sanityCheck('TIMERWRAP');
 
-  checkInvocations(t1, { init: 1, before: 2, after: 2, destroy: 1 },
+  checkInvocations(t1, { init: 1, before: 3, after: 3 },
```
ditto

Fishrock123 commented on test/parallel/test-handle-wrap-isrefed.js:
```diff
@@ -110,21 +110,4 @@ const dgram = require('dgram');
-  timer._handle.close(common.mustCall(() =>
-    strictEqual(timer._handle.hasRef(),
-                false, 'timer_wrap: not unrefed on close')));
-}
```
This test-case should be rewritten, not removed. `hasRef` here exists for diagnostic tooling purposes.

apapirovski replied:
I'm not certain how you envision this being rewritten? There's no longer a handle on the `timer`. We already have a test for TimerWrap separately.

Fishrock123 replied:
Fetching active handles should do, I think? Might be a bit to walk it to find the TimerWrap though. The other option is to export the lists object from `internal/timers` and then import that here.

apapirovski replied:
Relevant: my next PR removes TimerWrap altogether in favour of something more akin to how Immediates work. (It's already done, waiting on this one to land.) This test would be very short-lived...

Fishrock123 replied:
Won't that still use a uv handle with the possibility of being unrefed? If so, what is the difference - it should probably has a `hasRef()` anyways for the same reasons.

apapirovski replied:
It won't use the C++ HandleWrap anymore so it wouldn't be in getActiveHandles. Right now most diagnostics tools have to filter out TimerWrap manually because it's not very useful. For example, why-is-node-running filters it out from its reports.

(It'll work exactly like Immediates where the libuv internals aren't unnecessarily exposed.)

apapirovski replied:
Same as above, have a look at https://github.com/apapirovski/node/commit/d8f075da7b8129f59b1d537b6475394425b81c3e for more info.

Fishrock123 replied:
I still want this for diagnostic tooling if possible. It should be possible to tell if only unrefed timers exist.

(As relating to the second set of changes, I would imagine `Timer.now()` sees use, but we can discuss at that time.)

Fishrock123 commented on test/parallel/test-timers-next-tick.js:
```diff
@@ -0,0 +1,15 @@
+// This test documents an internal implementation detail of the Timers:
+// since timers of different durations are stored in separate lists,
+// a nextTick queue will clear after each list of timers. While this
+// behaviour is not documented it could be relied on by Node's users.
```
It was implied before - maybe it should be documented barring what I said earlier doesn't prevent the nextTick calls?

apapirovski replied:
Personally, I don't necessarily want to document it. We might want to have them run between individual timers in the future, honestly. That would be closer to spec & browsers.

Fishrock123 replied:
Hmm, fair enough.

Fishrock123 commented on test/parallel/test-timers-same-timeout-wrong-list-deleted.js:
```diff
@@ -30,41 +28,9 @@ const handle1 = setTimeout(common.mustCall(function() {
-function getActiveTimers() {
-  const activeHandles = process._getActiveHandles();
-  return activeHandles.filter((handle) => handle instanceof Timer);
-}
```
Maybe internalTimers should expose the list objects so that this test actually tests itself properly...?

I would say the case is theoretically still relevant, in concept at least.

apapirovski replied:
It should still test itself via behaviour, even without checking the handles. Unless I'm completely misunderstanding it?

Fishrock123 replied:
Yeah, I think you're right. The original was a bit more through but I'm also not convinced it is necessary.

apapirovski commented on lib/internal/priority_queue.js:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  peek() {
+    return this[kHeap][1];
```
It makes the math slightly more complicated to start binary heaps at 0, so it's unused.

apapirovski commented on lib/timers.js:
```diff
@@ -126,36 +122,59 @@ const kRefed = Symbol('refed');
-// timers within (or creation of a new list).
-// However, these operations combined have shown to be trivial in comparison to
-// other alternative timers architectures.
+// The PriorityQueue which manages the order of expiring Timeouts and object
```
@Fishrock123 yes, it is. I'll add a note.

apapirovski commented:
Also, just wanted to say thanks for the detailed review @Fishrock123 — I'm updating everything right now. Really appreciate the thorough look!

apapirovski commented:
Now that CI isn't so red, another run: https://ci.nodejs.org/job/node-test-pull-request/14980/

@Fishrock123 this is ready for another review when you have a moment. I think I addressed all the comments?

TimothyGu commented:
New benchmark CI: https://ci.nodejs.org/view/All/job/benchmark-node-micro-benchmarks/198/

BridgeAR commented:
Oh, I actually started a benchmark before and forgot to add a the link:

https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/196/console
https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/197/console

I reran the first one with 120 runs because the variance was a bit high.

BridgeAR commented:
I think it is safe to say that this will not regress the performance (120 runs of the latest code https://github.com/nodejs/node/pull/20555/commits/2284fd1e71809c11a8432650709a9a45ab636ed1).

```
                                                              confidence improvement accuracy (*)    (**)   (***)
 timers/timers-breadth.js n=5000000                                  ***     27.87 %       ±3.22%  ±4.25%  ±5.45%
 timers/timers-cancel-pooled.js n=5000000                            ***     23.01 %       ±7.32%  ±9.65% ±12.39%
 timers/timers-cancel-unpooled.js direction='end' n=1000000          ***     62.16 %      ±11.51% ±15.20% ±19.56%
 timers/timers-cancel-unpooled.js direction='start' n=1000000        ***     11.86 %       ±5.28%  ±6.96%  ±8.94%
 timers/timers-depth.js n=1000                                       ***     -0.50 %       ±0.16%  ±0.22%  ±0.28%
 timers/timers-insert-pooled.js n=5000000                             **     -2.54 %       ±1.89%  ±2.49%  ±3.20%
 timers/timers-insert-unpooled.js direction='end' n=1000000            *      2.24 %       ±1.82%  ±2.40%  ±3.09%
 timers/timers-insert-unpooled.js direction='start' n=1000000        ***    105.25 %       ±4.18%  ±5.52%  ±7.09%
 timers/timers-timeout-pooled.js n=10000000                                   1.84 %       ±2.90%  ±3.82%  ±4.90%
 timers/timers-timeout-unpooled.js n=1000000                         ***    736.53 %      ±25.97% ±34.32% ±44.23%
```

Fishrock123 commented:
I'll try to look monday or tues before I'm away - we should still rename the benchmarks though, since unpooled timers no longer actually exist.

TimothyGu commented:
@BridgeAR I saw! Those runs were a bit outdated due to recent changes, hence “new” benchmark ci :)

BridgeAR commented:
@TimothyGu I triggered them after the latest commit?

apapirovski commented:
@Fishrock123 I think only the `timeout-unpooled` doesn't have a valid name anymore? The rest are still unpooled since they create a new TimersList and require adding that to the PriorityQueue. The lack of a handle shouldn't make a difference to the naming. The architecture is still essentially the same, just implemented in JS.

TimothyGu commented:
@BridgeAR oh oops, never mind then :)

Fishrock123 commented:
I still think it would be safer (and technically correct) to have it under `semver-major` though. :/

apapirovski commented:
@Fishrock123 That's fine by me, it's already marked as such and I had no intention of changing that.

I've made the request change and this should now be ready. Please have another look when you have a moment.

CI: https://ci.nodejs.org/job/node-test-pull-request/15003/

Fishrock123 commented:
One last comment / discussion before I ✅

This makes it difficult to know if an individual timeout is unrefed (again, diagnostic purposes, and yes we do actually use that). Should `kRef` be public api?

apapirovski commented:
Landed in a5aad244b1...23a56e0c28. Thank you so much to everyone who took the time to review or even have a look at this! Onto the next related PRs.

Trott commented:
This appears to be the cause of https://github.com/nodejs/node/issues/20907. PR to revert: https://github.com/nodejs/node/pull/20919
